rust/tw_bitcoin/src/ordinals.rs

```diff
@@ -0,0 +1,162 @@
+    pub fn new(
+        mime: &[u8],
+        data: &[u8],
+        recipient: Recipient<PublicKey>,
+    ) -> Result<OrdinalsInscription> {
+        let envelope = create_envelope(mime, data, recipient.public_key())?;
+        // TODO: In which cases is this `false`?
+        let merkle_root = envelope.spend_info.merkle_root().unwrap();
+
+        Ok(OrdinalsInscription {
+            envelope,
+            recipient: Recipient::<TaprootScript>::from_pubkey_recipient(recipient, merkle_root),
+        })
+    }
+    pub fn recipient(&self) -> &Recipient<TaprootScript> {
+        &self.recipient
+    }
+}
+
+/// Creates an [Ordinals Inscription](https://docs.ordinals.com/inscriptions.html).
+/// This function is used for two purposes:
+///
+/// 1. It creates the spending condition for the given `internal_key`. This
+///    associates the public key of the recipient with the Merkle root of the
+///    Inscription on-chain, but it does not actually reveal the script to
+///    anyone ("commit stage").
+/// 2. The same function can then be used by the spender/claimer to actually
+///    transfer the Inscripion by sending a transaction with the Inscription
+///    script in the Witness ("reveal stage").
+///
+/// Do note that the `internal_key` can be different for each stage, but it
+/// could also be the same entity. Stage one, the `internal_key` is the
+/// recipient. Stage two, the `internal_key` is the claimer of the transaction
+/// (where the Inscription script is available in the Witness).
+fn create_envelope(mime: &[u8], data: &[u8], internal_key: PublicKey) -> Result<TaprootProgram> {
+    use bitcoin::opcodes::all::*;
+    use bitcoin::opcodes::*;
+
+    // TODO: Check overflow
+    let (op_push, size_buf) = get_op_push(mime.len() as u32)?;
+
+    // Prepare content-type buffer.
+    let mut mime_buf = PushBytesBuf::new();
+
+    // For any sizes below 75, we use encode as `OP_PUSHBYTES[0|1|2|...|75]`.
+    // Fany any sized above 75, we use encode as `OP_PUSHDATA[1|2|3|4] <SIZE_BUF>`.
+    mime_buf.push(op_push.to_u8()).unwrap();
+    if let Some(size_buf) = size_buf {
+        mime_buf.extend_from_slice(&size_buf).unwrap();
+    }
+    mime_buf.extend_from_slice(mime).unwrap();
+
+    // Prepare data buffer.
+    let mut data_buf = PushBytesBuf::new();
+    let (op_push, size_buf) = get_op_push(data.len() as u32)?;
+    data_buf.push(op_push.to_u8()).unwrap();
+    if let Some(size_buf) = size_buf {
+        data_buf.extend_from_slice(&size_buf).unwrap();
+    }
+    data_buf.extend_from_slice(data).unwrap();
+
+    // Create an Ordinals Inscription.
+    let script = ScriptBuf::builder()
+        .push_opcode(OP_FALSE)
+        .push_opcode(OP_IF)
+        .push_opcode(OP_PUSHBYTES_3)
+        .push_slice(b"ord")
+        .push_opcode(OP_PUSHNUM_1)
+        .push_slice(mime_buf)
+        .push_opcode(OP_PUSHBYTES_0)
+        .push_slice(data_buf)
+        .push_opcode(OP_ENDIF)
+        .into_script();
+
+    dbg!(&script);
```

Comment:
We should avoid printing anything to std.
So just not to forget when this PR is r2r:
https://doc.rust-lang.org/std/macro.dbg.html
> Note that the macro is intended as a debugging tool and therefore you should avoid having uses of it in version control for long periods (other than in tests and similar). Debug output from production code is better done with other facilities such as the [debug!](https://docs.rs/log/*/log/macro.debug.html) macro from the [log](https://crates.io/crates/log) crate.

Reply:
Yep, forgot to remove this :)